R/integrate_sces.R

```diff
@@ -0,0 +1,159 @@
+# This function can perform the crux integration based on the given integration method
+
+
+#' Integrate a combined set of SingleCellExperiment objects using a specified
+#'  integration method. The final SCE object will contain an additional reducedDim
+#'  entitled `{integration_method}_PCA`. All original SCE assays are retained.
+#'  `fastMNN` integration uses all default setting, but additional parameters can be supplied.
+#'  `harmony` integration uses existing PCs, but additional parameters can be supplied.
+#'
+#' @param combined_sce A combined SCE object as prepared by `scpcaTools::merge_sce_list()`.
+#' @param integration_method The integration method to use. One of `fastMNN` or
+#'  `harmony` (case-sensitive)
+#' @param covariate_cols A vector of covariates (e.g. the batches) to consider during integration.
+#'   For both `fastMNN` and `harmony`, this should be SCE column indicating batches.
+#'   For `harmony` specifically, additionally SCE columns may be supplied.
+#'   Default is `c("sample")`.
+#' @param return_corrected_expression A boolean indicating whether corrected expression
+#'   values determined by the given integration method should be included in the
+#'   integrated SCE object. Note that `harmony` does not calculate corrected expression,
+#'   so this argument is ignored for this integration method. Default is `TRUE`.
+#' @param seed Random seed to set for integration. This is only set if the value is not `NULL`.
+#' @param ... Any additional parameters to be passed to the given integration method
+#'
+#' @return An update combined SCE object containing a new reduced dimension named
+#'   `{integration_method}_PCA` containing the corrected PCs. In the case of `fastMNN`
+#'   integration, the SCE also includes a corrected expression assay (`fastMNN_corrected`)
+#'
+#' @import SingleCellExperiment
+#'
+#' @export
+integrate_sces <- function(combined_sce,
+                           integration_method = c("fastMNN", "harmony"),
+                           covariate_cols = "sample",
+                           return_corrected_expression = TRUE,
+                           seed = NULL,
+                           ...) {
+
+  # Set seed
+  set.seed(seed)
+
+  # make sure that input is a SingleCellExperiment
+  if(!is(combined_sce, "SingleCellExperiment")){
+    stop("The `combined_sce` must be a SingleCellExperiment object created with `scpcaTools::merge_sce_list().`")
+  }
```

Comment:
I think we may want two separate checks here. One check that this is a `SingleCellExperiment` object and then another check for the number of unique batches/samples that are provided. If there is only 1 sample in the SCE object then the error message you have here would be appropriate. 

Because of this, I think we do want a `batch_column` argument... I kind of like having a separate `covariate_cols` and `batch_column` argument because the `batch_column` is always required while the `covariate_cols` specifically refers to any extra covariates and is not a requirement.

Reply:
This is fine. I think I might make the additional argument just `covariate_columns` though, in case we add a future method that also allows covariates, and throw a warning if the method chosen does not support covariates. (I don't think that needs to be a fatal error).